domain/src/main/java/org/oppia/domain/oppialogger/crashlytics/ExceptionsController.kt

```diff
@@ -0,0 +1,174 @@
+import org.oppia.util.system.OppiaClock
+import javax.inject.Inject
+
+class ExceptionsController @Inject constructor(
+  private val exceptionLogger: ExceptionLogger,
+  private val dataProviders: DataProviders,
+  cacheStoreFactory: PersistentCacheStore.Factory,
+  private val consoleLogger: ConsoleLogger,
+  private val networkConnectionUtil: NetworkConnectionUtil,
+  private val oppiaClock: OppiaClock,
+  @ExceptionLogStorageCacheSize private val exceptionLogStorageCacheSize: Int
+) {
+  private val exceptionLogStore =
+    cacheStoreFactory.create("exception_logs", OppiaExceptionLogs.getDefaultInstance())
+
+  lateinit var exceptionCause: Throwable
+
+  fun logException(exception: Exception, timestamp: Long) {
+    uploadOrCacheExceptionLog(exception, timestamp)
+  }
+
+  /**
+   * Checks network connectivity of the device.
+   *
+   * Saves the [exceptionLog] to the [exceptionLogStore] in the absence of it.
+   * Uploads to remote service in the presence of it.
+   */
+  private fun uploadOrCacheExceptionLog(exception: Exception, timestamp: Long) {
+    when (networkConnectionUtil.getCurrentConnectionStatus()) {
+      NetworkConnectionUtil.ConnectionStatus.NONE ->
+        cacheExceptionLog(exceptionToExceptionLog(exception, timestamp))
+      else -> exceptionLogger.logException(exception)
+    }
+  }
+
+  /** Returns an [ExceptionLog] from an [exception]. */
+  private fun exceptionToExceptionLog(exception: Exception, timestamp: Long): ExceptionLog {
+    val exceptionLogBuilder = ExceptionLog.newBuilder()
+    val exceptionLogCauseBuilder = ExceptionLog.newBuilder()
+    exceptionLogBuilder.message = exception.message
+    exceptionLogBuilder.timestamp = timestamp
+    val stackTraceSize = exception.stackTrace.size
+    buildStackTraceElement(stackTraceSize, exceptionLogBuilder, exception)
+    exception.cause?.message?.let {
+      exceptionLogCauseBuilder.setMessage(it)
+    }
+    val causeStackTraceSize = exception.cause?.stackTrace?.size
+    if (causeStackTraceSize != null) {
+      buildStackTraceElement(causeStackTraceSize, exceptionLogCauseBuilder, exception)
+    }
+    exceptionLogBuilder.cause = exceptionLogCauseBuilder.build()
+    return exceptionLogBuilder.build()
+  }
+
+  /** Returns an [Exception] from an [exceptionLog]. */
+  private fun exceptionLogToException(exceptionLog: ExceptionLog): Exception {
+    val exceptionMessage = exceptionLog.message
+    exceptionCause = if (exceptionLog.cause.message != "") {
+      Throwable(exceptionLog.cause.message)
+    } else {
+      Throwable()
+    }
+    exceptionCause.stackTrace = createErrorStackTrace(exceptionLog.cause)
+    val exception = Exception(exceptionMessage, exceptionCause)
+    exception.stackTrace = createErrorStackTrace(exceptionLog)
+    return exception
+  }
+
+  /** Builds the [ExceptionLog.StackTraceElement] for an [exception] with relevant data and adds it to the [exceptionLogBuilder]. */
+  private fun buildStackTraceElement(
```

Comment:
You might be able to simplify the stack trace construction slightly:

```kotlin
exceptionLogBuilder.addAllStackTraceElement(exception.stackTrace.map(this::convertStackTraceElementToLog))

...

fun convertStackTraceElementToLog(stackTraceElement: StackTraceElement): ExceptionLog.StackTraceElement {
  return ExceptionLog.StackTraceElement.newBuilder()
      .setFileName(stackTraceElement.fileName)
      .setMethodName(stackTraceElement.methodName)
      .setLineNumber(stackTraceElement.lineNumber)
      .setDeclaringClass(stackTraceElement.className)
      .build()
}
```

Reply:
Yes, it helped. Thanks :)